### PR TITLE
feat: add node_modules/.bin binary files support to npm_translate_lock, npm_import and npm_link_package_store

### PR DESCRIPTION
**Pull request by gregmagolan**
This will be most useful for integration tests that want to run `node_modules/.bin` tools directly.

Only generating bash binaries for now. Windows can use sh_test.

For now the bin entries must be entered manually.

In the future, this field may be automatically populated by translate_lock_file from information in the pnpm lock file. That feature is currently blocked on https://github.com/pnpm/pnpm/issues/5131.

### CODE REVIEW

File: npm/private/npm_link_package_store.bzl [hunk 1]
```diff
@@ -34,6 +35,17 @@ If unset, the package name of the src npm_package_store is used.
 If set, takes precendance over the package name in the src npm_package_store.
 """,
     ),
+    "bins": attr.string_dict(
+        doc = """Dictionary of `node_modules/.bin` binary files to create mapped to their node entry points.
```
Comment (kormide) [inline npm/private/npm_link_package_store.bzl:39]: I think spelling out what the keys and values are here would be helpful. Or adding an sample.I wouldn't know what to put unless I saw an example.

File: npm/private/npm_link_package_store.bzl [hunk 2]
```diff
@@ -46,13 +58,25 @@ If set, takes precendance over the package name in the src npm_package_store.
     ),
 }
 
+_BIN_TMPL = """#!/bin/sh
+basedir=$(dirname "$(echo "$0" | sed -e 's,\\\\,/,g')")
+exec node "$basedir/{bin_path}" "$@"
+"""
+
```
Comment (kormide) [inline npm/private/npm_link_package_store.bzl:65]: Unrelated to this PR, but do we need a windows `.cmd` equivalent for these like I did [here](https://github.com/angular/angular/pull/47135/files#diff-2bb203c749eb393e1596af31197ef437a63f87fd6ac87dc6222b89f70fee37cfR239-R243)?

File: npm/private/npm_translate_lock.bzl [hunk 3]
```diff
@@ -164,21 +164,50 @@ _ATTRS = {
...
+
+        Binary files to create are declared as key/value pairs of the form "bin_name=bin_path".
+
+        For example:
+
+        ```
+        bins: {
+            "@foo/bar": ["foo=./foo.js", "bar=./bar.js"],
+        }
+        ```
```
Comment (kormide) [inline npm/private/npm_translate_lock.bzl:210]: Ah, I see you have an example here.

File: npm/private/npm_link_package_store.bzl (hunk 2 above)
Comment (gregmagolan) [reply npm/private/npm_link_package_store.bzl:65]: As a follow-up probably yes. For downstream work I'm going to use sh_test which works on Windows.

File: npm/private/test/BUILD.bazel [hunk 4]
```diff
@@ -48,3 +48,14 @@ filegroup(
...
     ],
 )
+
+sh_test(
+    name = "bin_test",
+    srcs = ["bin_test.sh"],
+    args = ["$(NODE_PATH)"],
+    data = [
+        ":node_modules/typescript",
+        "//:node_modules/typescript",
```
Comment (kormide) [inline npm/private/test/BUILD.bazel:58]: Is this testing something related to having two packages with conflicting bins? Perhaps a clarifying comment?

**Review by kormide — APPROVED** [review]
:taco:

File: npm/private/npm_link_package_store.bzl (hunk 1 above)
Comment (gregmagolan) [reply npm/private/npm_link_package_store.bzl:39]: Added more examples

File: npm/private/test/BUILD.bazel (hunk 4 above)
Comment (gregmagolan) [reply npm/private/test/BUILD.bazel]: Good call. Added comment.

File: npm/npm_import.bzl [hunk 5]
```diff
@@ -225,13 +237,35 @@ def npm_import(
...
+        bins: Dictionary of `node_modules/.bin` binary files to create mapped to their node entry points.
+
+            This is typically derived from the "bin" attribute in the package.json
+            file of the npm package being linked.
+
+            For example:
+
+            ```
+            bins: {
+                "foo": "./foo.js",
```
Comment (jbedard) [inline npm/npm_import.bzl:261]: Shouldn't the value be in an array? And have the `=`?

Is there a better package name then "foo" that clearly indicates that that is the package? Even just `pkg-foo`?

File: npm/private/npm_link_package_store.bzl [hunk 6]
```diff
@@ -76,6 +120,7 @@ def _impl(ctx):
     providers = [
         DefaultInfo(
             files = depset(files),
+            runfiles = ctx.runfiles(transitive_files),
```
Comment (jbedard) [inline npm/private/npm_link_package_store.bzl:123]: This was missing before? Is that related to this PR?

File: npm/npm_import.bzl (hunk 5 above)
Comment (jbedard) [reply npm/npm_import.bzl:261]: Oh I see, this is the `npm_import`, and it does align with the `npm_translate_lock` 👍

Comment (gregmagolan) [reply npm/npm_import.bzl:261]: Yeah, unfortunately there is no `string_list_dict_dict` so the equals is a work-around in npm_translate_lock.

Comment (gregmagolan) [reply npm/npm_import.bzl:261]: Tho I might do the translate from string_list_dict_dict in wrapper macro instead so the API is consistent

File: npm/private/npm_link_package_store.bzl (hunk 6 above)
Comment (gregmagolan) [reply npm/private/npm_link_package_store.bzl:123]: It related to the sh_test added. It lets you pass a linked npm package to the data of a generic rule, such as sh_test, and get those files in the runfiles.

Comment (jbedard) [reply npm/private/npm_link_package_store.bzl:123]: That seems like a fix/feature on its own that could land separately? Or just it's own commit 🤷 

Is there an easy test we can add for this?

Comment (gregmagolan) [reply npm/private/npm_link_package_store.bzl:123]: I'll add a simple sh_test and move it to it own commit

File: npm/private/npm_translate_lock.bzl [hunk 7]
```diff
@@ -6,203 +6,26 @@ load(":utils.bzl", "utils")
...
-        This points to a `.bazelignore` file to verify that all nested node_modules directories
-        pnpm will create are listed.
-
-        See https://github.com/bazelbuild/bazel/issues/8106
-        """,
-    ),
-    "warn_on_unqualified_tarball_url": attr.bool(
-        default = True,
-    ),
+    "pnpm_lock": attr.label(),
```
Comment (jbedard) [inline npm/private/npm_translate_lock.bzl:10]: This means bazel has no reference to the label docs?

File: npm/npm_import.bzl (hunk 5 above)
Comment (gregmagolan) [reply npm/npm_import.bzl:518]: ✅

File: npm/private/npm_translate_lock.bzl (hunk 7 above)
Comment (gregmagolan) [reply npm/private/npm_translate_lock.bzl:10]: Yeah. The docs have been hoisted to the macro now. Same as npm_import already does.

Comment (jbedard) [reply npm/private/npm_translate_lock.bzl:10]: Bazel no longer knows of the doc strings though. Is that ok? Does bazel use or provide those in any way?

Comment (gregmagolan) [reply npm/private/npm_translate_lock.bzl:10]: stardocs docs are still generated correct from the macro docstring summary and Args. Other than that the docstrings are not load bearing anywhere.

**Comment by gregmagolan** [comment]
Will be really nice when the pnpm lock file has bin entries and we can generate these automatically

File: npm/npm_import.bzl [hunk 8]
```diff
@@ -30,23 +30,243 @@ load("//npm/private:versions.bzl", "PNPM_VERSIONS")
...
+            For example:
+
+            ```
+            lifecycle_hooks_execution_requirements: {
+                "*": ["requires-network"],
+                "@foo/bar": ["no-sandbox"],
+            }
+            ```
+
+        bins: Binary files to create in `node_modules/.bin` for packages in this lock file.
```
Comment (jbedard) [inline npm/npm_import.bzl:234]: Is it worth specifying that the paths are relative to the package directory and not node_modules or node_modules/.bin? Maybe that's obvious enough 🤷

File: npm/npm_import.bzl [hunk 9]
```diff
@@ -57,11 +277,39 @@ def npm_translate_lock(name, npm_package_lock = None, yarn_lock = None, pnpm_ver
...
-    npm_translate_lock_rule(
+    # convert bins to a string_list_dict to satisfy attr type in repository rule
+    bins_string_list_dict = {}
+    if type(bins) != "dict":
+        fail("Expected bins to be a dict")
+    for key, value in bins.items():
+        if type(value) != "dict":
+            fail("Expected values in bins to be a dicts")
+        for value_key, value_value in value.items():
+            if key not in bins_string_list_dict:
```
Comment (jbedard) [inline npm/npm_import.bzl:288]: This could be up a line outside the loop

File: npm/npm_import.bzl [hunk 10]
```diff
@@ -30,23 +30,243 @@ load("//npm/private:versions.bzl", "PNPM_VERSIONS")
...
-npm_translate_lock_rule = repository_rule(
-    doc = _npm_translate_lock_lib.doc,
+_npm_translate_lock = repository_rule(
     implementation = _npm_translate_lock_lib.implementation,
     attrs = _npm_translate_lock_lib.attrs,
 )
 
 LATEST_PNPM_VERSION = PNPM_VERSIONS.keys()[-1]
 
-def npm_translate_lock(name, npm_package_lock = None, yarn_lock = None, pnpm_version = LATEST_PNPM_VERSION, **kwargs):
```
Comment (jbedard) [inline npm/npm_import.bzl:41]: Does this mean things like `tags` and `testonly` can no longer be specified? Is that ok or intentional?

**Review by jbedard — APPROVED** [review]

File: npm/npm_import.bzl (hunk 8 above)
Comment (gregmagolan) [reply npm/npm_import.bzl:234]: Should be obvious enough; it is also mentioned that the values are derived from the package.json of the npm package.

File: npm/npm_import.bzl (hunk 9 above)
Comment (gregmagolan) [reply npm/npm_import.bzl:288]: Nice catch

File: npm/npm_import.bzl (hunk 10 above)
Comment (gregmagolan) [reply npm/npm_import.bzl:41]: They never could in the first place since these are repository rules :)

Comment (jbedard) [reply npm/npm_import.bzl:41]: Perfect 👍
